aip/auth/4114.md

```diff
@@ -0,0 +1,93 @@
+---
+id: 4114
+scope: auth
+state: approved
+created: 2020-07-28
+---
+
+# Device Certificate Authentication via Mutual TLS
+
+Mutual TLS (a.k.a mTLS) authentication enables authentication of both client and server endpoints in a TLS handshake, and offers the solution to enable Context Aware Access for GCP enterprise customers. With Context Aware Access, the transport context considers the client certificate’s state in the transport (TLS) to make authorization decisions. The use of client certificates provides a stronger indication of the access originating from a trusted device. This allows customers to restrict access from untrusted devices.
+
+**Note:** Because this AIP describes guidance and requirements in a
+language-neutral way, it uses generic terminology which may be imprecise or
+inappropriate in certain languages or environments.
+
+## Guidance
+
+This section describes the general guidance of supporting Device Certificate
+Authentication (a.k.a DCA) via mTLS.
+
+### Application Default Credentials for DCA
+
+Users **should** enable DCA through [ADC][0] instead of manual configuration
+via client options.
+
+There are two aspects of ADC for DCA. The first is the automatic procurement of
+a device certificate. The second is the automatic switching of the service
+endpoint to the mTLS version of the service endpoint. Clients **must** support
+both of these aspects when supporting ADC for DCA.
+
+### Expected Behavior
+
+Client support for DCA **must** give priority to user overrides specified
+via client options. The following decision tree **should** be used:
+
+1.  If user specifies both device certificate and endpoint override via client
+    options, use them as is.
+1.  If user does not specify device certificate, attempt to procure and use
+    a default device certificate.
+1.  If user does not specify endpoint override, use the default mTLS endpoint if
+    a device certificate is available and the default regular endpoint otherwise.
+
+Implications of the above logic:
+
+1.  If user specifies a non-mTLS endpoint override but a device certificate is
+    available, pass along the certificate anyway and let the server decide what to do.
+1.  If user specifies an mTLS endpoint override but device certificate is not
+    available, do not fail-fast, but let server return error when connecting.
+
+The above behavior avoids introducing client-side logic that parses whether the
+endpoint override is an mTLS url, since the url pattern may change at anytime.
+
+### Obtaining the Default mTLS Endpoint
+
+The default mTLS endpoint for a service **should** be read from the Discovery
+Document field "mtlsRootUrl" instead of generated via regex patterns.
+
+### Obtaining the Default Device Certificate
+
+The default device certicate **should** be procured using the
```

Comment:
We need more details here, like which file to read, what the expected content of the file and what the key looks like

Reply:
done